### PR TITLE
feat(storage): add `StateTable::iter_with_pk_range` for `DynamicFilter`

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Add iter_with_pk_range for a given vnode

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Tracking
https://github.com/risingwavelabs/risingwave/issues/3419

### CODE REVIEW

File: src/storage/src/table/streaming_table/state_table.rs [hunk 1]
```diff
@@ -783,39 +850,48 @@ impl<S: StateStore> StateTable<S> {
         // If this assertion fails, then something must be wrong with the operator implementation or
         // the distribution derivation from the optimizer.
         let vnode = self.compute_vnode(pk_prefix).to_be_bytes();
-        let encoded_key_range_with_vnode = prefixed_range(encoded_key_range, &vnode);
+        let encoded_key_range_with_vnode = prefixed_range(encoded_key_range.clone(), &vnode);
```
Comment (wcy-fdu) [inline src/storage/src/table/streaming_table/state_table.rs:853]: remove this `clone()`?

File: src/storage/src/table/streaming_table/state_table.rs [hunk 2]
```diff
@@ -783,39 +850,48 @@ impl<S: StateStore> StateTable<S> {
...
+
+    async fn iter_inner(
+        &self,
+        key_range: (Bound<Vec<u8>>, Bound<Vec<u8>>),
+        prefix_hint: Option<Vec<u8>>,
+        epoch: u64,
+    ) -> StorageResult<(MemTableIter<'_>, StorageIterInner<S>)> {
         // Mem table iterator.
-        let mem_table_iter = self.mem_table.iter(encoded_key_range_with_vnode.clone());
+        let mem_table_iter = self.mem_table.iter(key_range.clone());
```
Comment (wcy-fdu) [inline src/storage/src/table/streaming_table/state_table.rs:883]: ditto.

File: src/storage/hummock_sdk/src/key.rs [hunk 3]
```diff
@@ -178,6 +178,18 @@ pub fn end_bound_of_prefix(prefix: &[u8]) -> Bound<Vec<u8>> {
...
 
+/// Get the start bound of the given `prefix` when it is excluded from the range.
+pub fn start_bound_of_excluded_prefix(prefix: &[u8]) -> Bound<Vec<u8>> {
+    if let Some((s, e)) = next_key_no_alloc(prefix) {
+        let mut res = Vec::with_capacity(s.len() + 1);
+        res.extend_from_slice(s);
+        res.push(e);
+        Included(res)
+    } else {
+        panic!() // We cannot deal with the case where the prefix is the maximum value
```
Comment (wcy-fdu) [inline src/storage/hummock_sdk/src/key.rs:189]:
```suggestion
        panic!("the prefix is the maximum value") 
```

File: src/storage/src/table/streaming_table/state_table.rs [hunk 4]
```diff
@@ -723,8 +726,70 @@ impl<S: StateStore> StateTable<S> {
...
+    }
+
+    /// This function scans rows from the relational table with specific `pk_prefix`.
+    pub async fn iter_with_pk_range<'a>(
+        &'a self,
+        pk_range: &'a (Bound<Row>, Bound<Row>),
+        // Optional vnode that returns an iterator only over the given range under that vnode.
+        // For now, we require this parameter, and will panic. In the future, when `None`, we can
+        // iterate over each vnode that the `StateTable` owns.
+        vnode: Option<u8>,
```
Comment (st1page) [inline src/storage/src/table/streaming_table/state_table.rs:748]:
```suggestion
        vnode: u8,
```
I think there will be no stream executor scanning data in multiple vnode at all. so maybe we can force the streaming executor give the vnode.

**Review by st1page — COMMENTED** [review]
rest LGTM

**Review by st1page — APPROVED** [review]
LGTM

**Pull request merged**